### PR TITLE
Create api for cancelling a recurring subscription #45

### LINKED ISSUE #45: Create API for cancelling a recurring subscription
1. Don't remove subscription records.
2. Recurring subscriptions ONLY.
3. Add `canceledDate` field.

### PR DESCRIPTION
**Pull request by andy-sam**
fixes #45

### CODE REVIEW

File: src/main/java/com/codesarge/banking/controllers/SubscriptionController.java [hunk 1]
```diff
@@ -35,4 +38,18 @@ public SubscriptionController(
         }
         return ResponseEntity.ok(responseDto);
     }
+
+    @DeleteMapping("bank/services/subscriptions/{id}")
+    @ResponseBody
+    public ResponseEntity<SubscriptionDto> cencelSubscription(
```
Comment (alex-anakin) [inline src/main/java/com/codesarge/banking/controllers/SubscriptionController.java:44]: censel -> cancel

File: src/main/java/com/codesarge/banking/repositories/SubscriptionRepository.java [hunk 2]
```diff
@@ -8,4 +8,12 @@
 
 public interface SubscriptionRepository extends JpaRepository<Subscription, Long> {
     List<Subscription> findSubscriptionsByBankServicePeriodicity(Periodicity periodicity);
+
+    List<Subscription> findSubscriptionsByAccountId(Long accountId);
```
Comment (alex-anakin) [inline src/main/java/com/codesarge/banking/repositories/SubscriptionRepository.java:12]: please remove

File: src/main/java/com/codesarge/banking/service/SubscriptionService.java [hunk 3]
```diff
@@ -46,4 +52,23 @@ public SubscriptionDto makeSubscription(SubscriptionRequestDto subscriptionReque
         accountService.withdrawMoney(account, bankService.getPrice());
         return savedSubscriptionDto;
     }
+
+    public SubscriptionDto deleteSubscription(Long subscriptionId) {
+        Subscription subscription = subscriptionRepository.getById(subscriptionId);
```
Comment (alex-anakin) [inline src/main/java/com/codesarge/banking/service/SubscriptionService.java:57]: wrap this call in try/catch
in `catch` block handle `Exception` and return null

File: src/main/java/com/codesarge/banking/service/SubscriptionService.java [hunk 4]
```diff
@@ -46,4 +52,23 @@ public SubscriptionDto makeSubscription(SubscriptionRequestDto subscriptionReque
...
+        Long accountId = subscription.getAccount().getId();
+        Long bankServiceId = subscription.getBankService().getId();
+
+        SubscriptionDto canceledSubscriptionDto = null;
+
+        if (subscription.getBankService().getPeriodicity().equals(Periodicity.WEEKLY) ||
```
Comment (alex-anakin) [inline src/main/java/com/codesarge/banking/service/SubscriptionService.java:63]: extract a variable `subscription.getBankService().getPeriodicity()`

File: src/main/java/com/codesarge/banking/service/SubscriptionService.java [hunk 5]
```diff
@@ -46,4 +52,23 @@ public SubscriptionDto makeSubscription(SubscriptionRequestDto subscriptionReque
...
+                subscription.getBankService().getPeriodicity()
+                        .equals(Periodicity.MONTHLY)) {
+            if (subscription.getCanceledDate() == null) {
```
Comment (alex-anakin) [inline src/main/java/com/codesarge/banking/service/SubscriptionService.java:66]: combine in one `if` statement with previous conditions

File: src/main/java/com/codesarge/banking/service/SubscriptionService.java [hunk 6]
```diff
@@ -46,4 +52,23 @@ public SubscriptionDto makeSubscription(SubscriptionRequestDto subscriptionReque
...
+                subscription.setCanceledDate(LocalDateTime.now());
+                canceledSubscriptionDto = new SubscriptionDto(subscription);
+                subscriptionRepository.save(subscription);
```
Comment (alex-anakin) [inline src/main/java/com/codesarge/banking/service/SubscriptionService.java:69]: move saving before creating a DTO

**Review by alex-anakin — CHANGES_REQUESTED** [review]

File: src/main/java/com/codesarge/banking/controllers/SubscriptionController.java (hunk 1 above)
Comment (andy-sam) [reply src/main/java/com/codesarge/banking/controllers/SubscriptionController.java:44]: done

File: src/main/java/com/codesarge/banking/repositories/SubscriptionRepository.java (hunk 2 above)
Comment (andy-sam) [reply src/main/java/com/codesarge/banking/repositories/SubscriptionRepository.java:12]: done

File: src/main/java/com/codesarge/banking/service/SubscriptionService.java (hunk 4 above)
Comment (andy-sam) [reply src/main/java/com/codesarge/banking/service/SubscriptionService.java:63]: done

File: src/main/java/com/codesarge/banking/service/SubscriptionService.java (hunk 5 above)
Comment (andy-sam) [reply src/main/java/com/codesarge/banking/service/SubscriptionService.java:66]: done

File: src/main/java/com/codesarge/banking/service/SubscriptionService.java (hunk 6 above)
Comment (andy-sam) [reply src/main/java/com/codesarge/banking/service/SubscriptionService.java:69]: done

File: src/main/java/com/codesarge/banking/service/SubscriptionService.java (hunk 3 above)
Comment (andy-sam) [reply src/main/java/com/codesarge/banking/service/SubscriptionService.java:57]: done

**Review by alex-anakin — APPROVED** [review]

**Pull request merged**